src/communication/RpcServer.cpp

```diff
@@ -10,3 +10,125 @@
 // SPDX-License-Identifier: Apache-2.0
 
 #include "up-cpp/communication/RpcServer.h"
+
+namespace uprotocol::communication {
+
+namespace Validator = datamodel::validator;
+
+RpcServer::RpcServer(std::shared_ptr<transport::UTransport> transport,
+                     const v1::UUri& method,
+                     std::optional<v1::UPayloadFormat> format,
+                     std::optional<std::chrono::milliseconds> ttl)
+    : transport_(std::move(transport)), ttl_(ttl) {
+	auto [valid, reason] = Validator::uri::isValidRpcMethod(method);
+	if (!valid) {
+		throw Validator::uri::InvalidUUri(
+		    "Method URI is not a valid URI |  " +
+		    std::string(Validator::uri::message(reason.value())));
+	}
+
+	if (format.has_value()) {
+		if (!UPayloadFormat_IsValid(format.value())) {
+			throw std::out_of_range("Invalid payload format");
+		} else {
+			expected_payload_format_ = std::move(format);
+		}
+	}
+}
+
+RpcServer::ServerOrStatus RpcServer::create(
+    std::shared_ptr<transport::UTransport> transport,
+    const v1::UUri& method_name, RpcCallback&& callback,
+    std::optional<v1::UPayloadFormat> payload_format,
+    std::optional<std::chrono::milliseconds> ttl) {
+	std::unique_ptr<RpcServer> server;
+	try {
+		server.reset(new RpcServer(std::move(transport), method_name,
+		                           std::move(payload_format), ttl));
+	} catch (const Validator::uri::InvalidUUri& e) {
+		v1::UStatus status;
+		status.set_code(v1::UCode::INVALID_ARGUMENT);
+		status.set_message(e.what());
+		return uprotocol::utils::Unexpected(status);
+	} catch (const std::out_of_range& e) {
+		v1::UStatus status;
+		status.set_code(v1::UCode::OUT_OF_RANGE);
+		status.set_message(e.what());
+		return uprotocol::utils::Unexpected(status);
+	} catch (const std::exception& e) {
+		v1::UStatus status;
+		status.set_code(v1::UCode::UNKNOWN);
+		status.set_message(e.what());
+		return uprotocol::utils::Unexpected(status);
+	}
```

Comment:
We don't use try/catch for flow control. If an exception was thrown, it is a panic that should be seen by the developer.

Instead, the only UStatus that should be returned by this function should be the one returned by `transport->registerListener()` (via `RpcServer::connect()`).

Also, don't use `new` directly. Instead, favor `std::make_unique`.

```suggestion
	auto server = std::make_unique<RpcServer>(transport, method_name, payload_format, ttl);
```

Reply:
yes originally U tried to use std::make_unique compiler throws error that it can't see constructor for this class as its private therefore using manual way which should be same 

 Compiler error:
 
```
 'uprotocol::communication::RpcServer::RpcServer(std::shared_ptr<uprotocol::transport::UTransport>, std::optional<uprotocol::v1::UPayloadFormat>, std::optional<std::chrono::duration<long int, std::ratio<1, 1000> > >)' is protected within this context
 1070 |     { return unique_ptr<_Tp>(new _Tp(std::forward<_Args>(__args)...)); }

```

Reply:
addressed thanks for suggestion